### DISCUSSION TITLE
SSR `redirect()` in appDir flashes unuthenticated content

### DISCUSSION THREAD

**Discussion by milovangudelj** [vercel/next.js#45746] [Help]
Hi, I'm migrating my website to use the new `app` directory and so far everything has worked flawlessly except for server side redirection.

I have a page that is protected by authentication using `next-auth`. If a visitor is unauthenticated it gets correctly redirected to the login page, but not before seeing a flash of unauthenticated content.

Previously this would not have happened since if there was no session data `getServerSideProps` would have returned a redirect object like so:
```jsx
import { getServerSession } from "next-auth";

export async function getServerSideProps(context) {
  const session = await getServerSession();

  if (!session) {
    return {
      redirect: {
        destination: '/login',
        permanent: false,
      },
    }
  }

  return {
    props: {}, // will be passed to the page component as props
  }
}
```

Now using the new `redirect()` function you can still see a flash of the protected page's layout, and the url goes, for example, from `/` to `/protected` and then to `/login`.

Am I missing something?

**Comment by leebenson** [comment]
Agreed, this is not ideal. I think fundamentally, the issue is that when a `redirect` is thrown within the `app` folder, it should short-circuit rendering and return a 302/307 at the server level _without_ returning any of the bootstrap or initial HTML.

Instead, a `200` is returned along with the full client-side logic. It's not until the client parses `redirect` that it does anything.

In summary -- this should be server driven, not client.

**Comment by leebenson** [comment]
As a follow-up to this -- it seems when using the Node.js runtime, a `307` _is_ returned.

When using `experimental-edge`, a 200 is returned.

**Comment by leebenson** [comment]
Opened https://github.com/vercel/next.js/issues/46437

**Comment by melanieseltzer** [comment]
Running into the same friction migrating my own app. Did you ever get this sorted out, @milovangudelj?

From some research/experimenting, it seems redirecting in [Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware) is the way (runs before the request is completed).

`next-auth` has a [middleware package](https://next-auth.js.org/configuration/nextjs#middleware) that does what we'd need (redirects unauthorized pages to the sign in page automatically). However, it only supports the jwt strategy which is a blocker for me because I'm using database :/

Apparently there's a [next-auth v5 release](https://github.com/nextauthjs/next-auth/pull/7443) on the horizon that's supposed to fix this. Seems like people are already using it with the [next-auth@experimental](https://github.com/nextauthjs/next-auth/pull/7443#issuecomment-1594929031) tag, so this will probably be my next step.

**Comment by robertmalicke** [comment]
Same issue here! Except I might say that I see a flash of layout.tsx and nothing from page.tsx which sounds like an improvement but it's pretty dang ugly watching layout.tsx with an empty page component flash!
